File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.py
import sys
from collections import deque
from collections import Counter


def solution(n, edge):
    adj = [[] for _ in range(n + 1)]
    distance = [sys.maxsize for _ in range(n + 1)]
    for a, b in edge:
        adj[a].append(b)
        adj[b].append(a)
    dq = deque()
    dq.appendleft([1, 0])
    distance[1] = 0
    while dq:
        idx, dis = dq.pop()
        for next_idx in adj[idx]:
            if distance[next_idx] > dis + 1:
                dq.appendleft([next_idx, dis + 1])
                distance[next_idx] = dis + 1
    distance = distance[1:]
    return Counter(distance).get(max(distance))

File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        Map<Integer, ArrayList<Integer>> adj = new HashMap<>();
        
        for (int i = 0; i <= n; i++){
            
            adj.put(i, new ArrayList<Integer>());
        }
        
        for (int[] e: edge){
            adj.get(e[0]).add(e[1]);
            adj.get(e[1]).add(e[0]);
        }
        
        Deque<Coord> dq = new LinkedList<>();
        int[] visited = new int[n + 1];
        
        for (int i = 2; i <= n; i++)
            visited[i] = n + 1;
        visited[0] = 0;
        visited[1] = 0;
        
        dq.add(new Coord(1, 0));
        
        while (!dq.isEmpty()){
            Coord cur = dq.poll();
            for (Integer next: adj.get(cur.a)){
                if (visited[next] > cur.b + 1){
                    visited[next] = cur.b + 1;
                    dq.add(new Coord(next, cur.b + 1));
                }
            }   
        }
        
        int cnt = 0;
        int maxValue = Arrays.stream(visited).max().getAsInt();
    
        
        // for (int i = 0; i <= n; i++){
        //     System.out.println(i + ", " + visited[i]);
        // }
        
        for (int distance: visited){
            if (distance == maxValue)
                cnt++;
        }   
        
        return cnt;
    }
}

class Coord{
    int a;
    int b;
    public Coord(int a, int b){
        this.a = a;
        this.b = b;
    }
}